Allow outputs functions to return const T&.

chrisdembia opened the pull request:
This addresses #824 for the Hackathon, and must be revisited afterwards. This change is in accord with what I discussed with @aseth1 and includes warnings in Doxygen to scare people away from using this variant of the Output.

aymanhab commented:
Looks fine but I'm not very familiar with the gymnastics involved, can you review and merge @klshrinidhi ? Thanks.

chrisdembia commented:
@aseth1 I added a comment to the code. I suggest we merge it before the tests pass, since I only added a comment.

I also added our discussion to #824 (the associated issue) to keep track of what we must do after the hackathon.

aymanhab commented:
Great, thanks @aseth1 & @chrisdembia

chrisdembia commented:
Thank you @aseth1
